M2/Macaulay2/packages/GKMManifolds.m2

```diff
@@ -0,0 +1,1462 @@
+    Y := tVariety(momentGraph(V,E,G.HTpt),X.charRing);
+    cellOrder(Y.momentGraph, subposet(P,V));
+    Y.cache.lieType = lieType X;
+    Y
+    )
+
+--------------------------------------------------------------------------------------------------
+--------------------------------< T-orbit closures of points >------------------------------------
+
+---------------------------< auxiliary functions for TOrbClosure >--------------------------------
+convertToNum = (n,L) -> (
+    return apply(toList L, v -> if v === unastrsk(v) then v else n + unastrsk v)
+    )
+
+revMat = M -> return matrix apply( reverse entries M, v-> reverse v)
+
+-- Takes in a mutable matrix and outputs the RREF with the identitiy block in the beginning
+rowRed = M -> (
+    Mat := matrix revMat(M) ;
+    return mutableMatrix revMat transpose gens gb image transpose Mat;
+    )
+
+
+
+--the T-equivariant K-class of a torus orbit closure of a point in a generalized flag variety
+--input:  X a tGeneralizedFlagVariety and MatLst a list of matrices, {M1,...,Mn}, that defines a point in X.
+--    	  For convenience we assume that the ranks of the M_i are distinct.    	
+--output: The tKClass of the closure of the orbit of the point corresponding to {M1,...,Mn}
+tOrbClosure = method()
+tOrbClosure(TVariety,List) := TKClass => (X,MatLst) -> ( 
+    if X.cache.?lieType then Typ := X.cache.lieType else (
+	 error "T-orbit closures are only implemented for Lie types"
+	 );
+    R := X.charRing;
+    m := numgens X.charRing;
+    x := symbol x;
+    S := QQ[x_0..x_(m-1)];
+    QS := frac(QQ[x_0..x_(m-1), Degrees => apply(m, i -> setIndicator(set {i},m))]);
+    rks :=  apply(first X.points, v -> #(elements v));
+    if not all(MatLst, v -> numcols v == (
+	    if Typ === "A" then m
+	    else if Typ === "B" then 2*m+1
+	    else 2*m
+	    )
+	) then error "the column size of the matrices are incorrect";
+    if not rks === apply(MatLst, v -> rank v) then (
+	error " the rank of the matrices are incorrect "
+	);
+    (if Typ === "A" then Gens := apply(gens QS, v -> v^(-1))
+	else if Typ  === "C" or Typ === "D" then Gens = apply(gens QS, v -> v^(-1)) | gens QS
+	else if Typ === "B" then Gens = apply(gens QS, v -> v^(-1)) | gens QS |{1}
+	);
+    Lst := apply(X.points, pt -> (
+	    degList := {};
+	    convertPt := apply(pt, v -> convertToNum(m,v));
+	    if any(#rks, v -> determinant (rowRed MatLst_v)_(convertPt_v) == 0) then 0_R else (
+		for j in toList(0..#rks-1) do (
+		    L := (reverse convertPt)#j;
+		    M := mutableMatrix sub((reverse MatLst)#j,QS);
+		    for i in toList(0..#L-1) do M = columnSwap(M,i,L#i);
+		    M = rowRed M;
+		    for i in reverse toList(0..#L-1) do M = columnSwap(M,i,L#i);
+		    for i in L do M = columnMult(M,i,0);
+		    for v in toList(0..#L-1) do M = rowMult(M,v,sub(Gens_(L_v)^(-1),ring M));
+		    for v in toList(0..#Gens-1) do M = columnMult(M,v,sub(Gens_v, ring M));
+		    degList = degList | apply(flatten entries M, v-> degree v);
```

Comment:
Another case of repeated concatenation, which might be slow.  Is one concatenation at the end possible?